proposed/project-file-simplification.md

```diff
@@ -0,0 +1,351 @@
+
+This information should be semantically available (possibly through MSBuild metadata) and also be helpful to humans perusing the file to see what's new.
+
+#### Changing the project file, retaining behavior
+
+One approach to the tooling would be to run a process that updates the current project to one that has identical behavior using the new `Base` by placing new properties into the project file. The user could then explore which of these to remove to enable new behavior. The user would start this process with a command like:
+
+```bash
+dotnet project upgrade 11.0.100
+```
+
+This would update the project file with the requested `Base` and new properties based on a _diff_ between the previous and resulting project properties.
+
+For example, if the 11.0.100 version of `Base` had two new enabled properties `UseFeatureA` and `UseFeatureB`, two new properties would be added to the project file when upgrading from a version 10 to version 11:
+
+```XML
+<Project Base="10.0.100" 
+         Type="Console"> 
+</Project> 
+```
+
+```XML
+<Project Base="11.0.100" 
+         Type="Console">
+    <PropertyGroup Label="These properties added on upgrade from 10.0.100 to 11.0.100"
+       <UseFeatureA>disable</UseFeatureA>
+       <UseFeatureB>disable</UseFeatureB>
+    </PropertyGroup>
+</Project> 
+```
+
+This approach is simple for a project file. It might be extended to complex repos by placing the `PropertyGroup` containing the new properties at the start of whichever file (project file, `Directory.Build.props` or `Import` file) the `Base` attribute appeared in. However, the command would probably be run in the same directory as the file containing the `Base` attribute and specify the file if there were multiple props files.
+
+#### Interactive upgrade
+
+|Command|Result|
+|-|-|
+|dotnet project upgrade 11.0.100 --whatif|List the properties that would be added, along with any inline comments|
+|dotnet project upgrade 11.0.100 --preserve|Actually run the upgrade, resulting in new properties to match previous behavior|
+|dotnet project upgrade 11.0.100 --no-preserve|Actually run the upgrade, without new properties|
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This introduces a new concept to .NET users, and some additional complexity in the overall MSBuild evaluation. Users would see this in areas such as understanding binlogs. We think the trade off is worthwhile because users get small clean project files the ability to control new warnings/breaking changes independent of both SDK version and TargetFramework.
+
+## Alternatives
+[alternatives]: #alternatives
+
+We could continue to tie new features to TFM. THe drawback is that users generally upgrade their TFM to keep their projects in support. While this avoids breaking changes/warnings occurring on tooling upgrade, it may either delay adopting desirable new features, or result in breaking changes/warnings happening when the user is just trying to stay in support.
+
+## Open questions
```

Comment:
How would this all work for other SDKs, say the iOS SDK?

For instance, given this project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net10.0-ios</TargetFramework>
  </PropertyGroup>
</Project>
```

How would this be in the new world? I'm guessing something like this:

```xml
<Project Base="10.0.100" 
         Type="iOS"> 
</Project> 
```

* Where is the fact that `Type=iOS` is a valid type declared? How does this translate to letting the user know they have to install the iOS SDK if it's not already installed?
* How can the iOS SDK provide their own set of defaults?
* How can these iOS defaults be versioned? The release schedule / versioning of the iOS SDK does not match the .NET SDK at all.

Reply:
Many similar scenarios will require multi-targeting, so it's a tad more complex. Especially because I think it's an error to declare both a `TargetFramework` and `TargetFrameworks` I added to the open question on multi-targeting.

Reply:
The Type can supply its own defaults.

For versioned defaults, we discussed looking for a file that next to the Base that defines Type specific defaults. Like:

- Base10.0.100.props
- iOS10.0.100.props

But, that didn't make it into the proposal. 

I need to review how workloads install as, at least for V1, we planned to restrict the location for Type and Base files to the .NET install repo. Hmmm.